### PR TITLE
THRIFT-5190: StringUtils haven't take `(offset + length) > bytes.length` into account

### PR DESCRIPTION
**Pull request by dugenkui03**
1. StringUtils haven't take `(offset + length) > bytes.length` into account;

2. `if ((offset | length | (offset + length) | (size - (offset + length))) < 0)` will faster than `>`

- [x]  [Apache Jira-THRIFT-5190](https://issues.apache.org/jira/browse/THRIFT-5190) 
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [ ] If your change does not involve any code, add ` [skip ci]` at the end of your pull request to free up build resources.

### CODE REVIEW

File: lib/java/src/org/apache/thrift/utils/StringUtils.java [hunk 1]
```diff
@@ -49,12 +49,10 @@ public static String bytesToHexString(byte[] bytes) {
    * @return hex string.
    */
   public static String bytesToHexString(byte[] bytes, int offset, int length) {
-    if (length < 0) {
-      throw new IllegalArgumentException("Negative length " + length);
-    }
-    if (offset < 0) {
-      throw new IndexOutOfBoundsException("Negative start offset " + offset);
+    if ((offset | length | (offset + length) | (bytes.length - (offset + length))) < 0) {
```
Comment (Jens-G) [inline lib/java/src/org/apache/thrift/utils/StringUtils.java:52]: To be honest,I personally have a problem with code like this where you need to read it three times to get it right (and I'm still not fully sure I "groked" it). I leave it to someone else to give a final review statement. +0 from me stince I won't be standing in the way, but I'm not really in favor of it.

Comment (Jens-G) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java:52]: A testcase for this change would be a great addition.

Comment (joaopedroantonio) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java:52]:
```suggestion
    if (offset < 0 || bytes.length < (offset + length)) {
```

And maybe we should keep the IllegalArgumentException when length is less than 0?

Comment (ctubbsii) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java:52]: I'm responding to the request on the mailing list for another pair of eyes on this PR.

I agree with @Jens-G regarding the readability... readability is far more important than an every-so-slight performance benefit. Obfuscated code is very hard to maintain, even with good test cases.

@joaopedroantonio 's suggestions are a great improvement, and preserves existing behavior, while adding the bounds checks when using offset and length.

A unit test case that covers the edge cases would be a nice sanity check for this method.

Comment (dugenkui03) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java:52]: > if ((offset | length | (offset + length) | (bytes.length - (offset + length))) < 0)

when length is less than 0, it will throw IllegalArgumentException, such as `int offset=2,length=-1,byteLength=10;`

Comment (dugenkui03) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java:52]: it seem that there will always be readability problem when using bit operation.

I do agree with that the readability of code is important. The code will be changed and squashed to a single commit.

Thanks very much.

Comment (joaopedroantonio) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java]: Maybe I'm being picky here but I think we should keep the IllegalArgumentException when `length < 0`, for two reasons:
1. The semantics of IndexOutOfBoundsException and the IllegalArgumentException are different and I think `length < 0` is a case of the latter.
2. Neither of those exceptions is checked and changing the type of exception when `length < 0` will not be explicit in compile time. I'd consider this a breaking change in the API.

WYT?

Comment (dugenkui03) [reply lib/java/src/org/apache/thrift/utils/StringUtils.java]: I haven't notice that I have removed IllegalArgumentException instead of IndexOutOfBoundsException, and I haven't notice that it will be a breaking change in the API.

great suggestion, thanks very much.

File: lib/java/test/org/apache/thrift/utils/TestStringUtils.java [hunk 2]
```diff
@@ -31,4 +31,27 @@ public void testToHexString() {
     Assert.assertEquals("EFAB92", StringUtils.bytesToHexString(bytes, 2, 3));
     Assert.assertNull(StringUtils.bytesToHexString(null));
   }
+
+  @Test
+  public void testToHexStringIndexOutOfBoundsExceptionExamine() {
+    byte[] bytes = {1, 2, 3, 4, 5};
+    try {
+      StringUtils.bytesToHexString(bytes, 0, -1);
+    } catch (Exception e) {
```
Comment (Jens-G) [inline lib/java/test/org/apache/thrift/utils/TestStringUtils.java:40]: The (hypothetical?) case when it would not throw is not covered. Same below.
